Add version command to mixer server cmd

douglas-reid opened the pull request:
This PR establishes a basic versioning for bazel builds to help
identify mixer builds and track compatibility across istio
components. It introduces a script to extract the date and git
commit SHA for a build. The script is used by bazel to write
status files which are then linked into the new version.go file
via the `linkstamp` rule of the BUILD file.

With these changes, a simple `bazel build //cmd/server:mixs` should
produce a binary that can be executed as follows to retrieve a
build id:

```shell
$ ./bazel-bin/cmd/server/mixs version
build: 2017-03-23-359f709
```
Addresses #424.

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/istio/mixer/448)
<!-- Reviewable:end -->

mandarjog commented on cmd/server/version.go:
```diff
@@ -0,0 +1,22 @@
+	BuildID = "nil"
+)
+
+func versionCmd() *cobra.Command {
```
I think we want BuildID to be available more generally, and not just in the command line.
It could be *used* by the version command. 

For example when we add a `/healtz`    or `/info` endpoint it needs to report the same info.
I suggest we move the `BuildID` to   `pkg/aspect. BuildID` or similar.

douglas-reid commented:
@mandarjog PTAL.  I've made the following changes:

- moved version stuff to //pkg/version
- updated scripts to include info on version (from tags) and clean/modified build status
- moved the version cmd from //cmd/server to //cmd/server/cmd

Now, with a basic bazel build, we can run the `mixs version` and get output like: `version: unknown (build: 2017-03-23-dd6d3ce, status: Clean)`

istio-testing commented:
Jenkins job mixer/presubmit passed

istio-testing commented:
Jenkins job manager/mixer-pr passed

istio-testing commented:
Jenkins job mixer/presubmit passed

geeknoid commented on cmd/server/cmd/version.go:
```diff
@@ -0,0 +1,32 @@
+func versionCmd(outf outFn) *cobra.Command {
+	versionCmd := cobra.Command{
+		Use:   "version",
+		Short: "Prints out build version information for Mixer",
```
We generally use mixer as a common noun, so that should be "for the mixer".

douglas-reid replied:
done.

geeknoid commented on cmd/server/cmd/BUILD:
```diff
@@ -8,6 +8,7 @@ go_library(
         "inventory.go",
         "root.go",
         "server.go",
+        "version.go",
```
Can we add the exact same version command to mixc so we can tell whether it's compatible with mixs?

douglas-reid replied:
I've now moved the version command into a shared package (along with OutFn and ErrorFn).

geeknoid commented on cmd/server/cmd/version.go:
```diff
@@ -0,0 +1,32 @@
+			outf("%s\n", version.Info)
+		},
+	}
+	return &versionCmd
```
Can just use:

return &cobra.Command{...

douglas-reid replied:
done.

douglas-reid commented:
PTAL.

Changes:
- new //cmd/shared package
- version command added to mixc

istio-testing commented:
Jenkins job mixer/presubmit passed

geeknoid commented on cmd/shared/shared.go:
```diff
@@ -0,0 +1,40 @@
+func VersionCmd(outf OutFn) *cobra.Command {
+	return &cobra.Command{
+		Use:   "version",
+		Short: "Prints out build version information for the mixer",
```
This is no longer really accurate given the fact this is now shared code...

douglas-reid replied:
Updated.

geeknoid commented on cmd/shared/shared.go:
```diff
@@ -0,0 +1,40 @@
+// OutFn is a function used for normal output.
+type OutFn func(format string, a ...interface{})
+
+// ErrorFn is a function used for error output.
```
Worth mentioning that calls to this function do not return.

douglas-reid replied:
Addressed in a different manner in latest commit.

geeknoid commented on cmd/shared/shared.go:
```diff
@@ -0,0 +1,40 @@
+)
+
+// OutFn is a function used for normal output.
+type OutFn func(format string, a ...interface{})
```
Given this has been elevated to some shared code rather than being an implementation detail of a command, I wonder if this stuff would be better if we gave the commands an io.Writer instead of a function pointer? It would compose better and would still provide the same flexibility to capture all the output in tests...

douglas-reid replied:
Perhaps, but then I think we want to have something that matches the fatal signature. My latest commit attacks this in a different manner.

douglas-reid commented:
PTAL.

Changes:
- Refactors OutFn and ErrorFn to shared.FormatFn.
- Added shared.Printf and shared.Fatalf impls

geeknoid commented:
LGTM

istio-testing commented:
Jenkins job mixer/presubmit passed
